File: scripts/import_vehicle_master.dart
#!/usr/bin/env dart
// ignore_for_file: avoid_print
/// Vehicle Master CSV → Firestore Import Script
///
/// Usage:
///   dart scripts/import_vehicle_master.dart [--csv data/vehicle_masters.csv] [--dry-run]
///
/// Options:
///   --csv <path>   CSV file path (default: data/vehicle_masters.csv)
///   --dry-run      Preview changes without writing to Firestore
///   --clear        Clear existing Firestore data before import
///
/// CSV Format:
///   type,id,parent_id,name,name_en,body_type,production_start_year,production_end_year,display_order,country
///
/// Requirements:
///   - Firebase service account key at: .env or GOOGLE_APPLICATION_CREDENTIALS
///   - Run from project root
///
/// When to run:
///   - Initial setup: dart scripts/import_vehicle_master.dart
///   - New model year: add row to CSV, then re-run
///   - New maker added: add maker + models to CSV, then re-run

import 'dart:io';

// NOTE: This script uses firebase_admin SDK (Node.js) via shell, OR
// can be adapted to use Firestore REST API directly.
// For Flutter projects, the recommended approach is:
//
// Option A: Use FlutterFire CLI + Dart Firebase Admin (dart_firebase_admin)
// Option B: Use firebase CLI: firebase firestore:import
// Option C: Run as Flutter integration test that calls VehicleMasterService.seedMasterData()
//
// This script implements Option C (safest for Flutter projects):
// It generates a Dart test file that seeds data using the existing VehicleMasterService.

void main(List<String> args) async {
  final csvPath = _getArg(args, '--csv') ?? 'data/vehicle_masters.csv';
  final isDryRun = args.contains('--dry-run');
  final doClear = args.contains('--clear');

  print('=== Trust Car Platform: Vehicle Master Import ===');
  print('CSV: $csvPath');
  if (isDryRun) print('[DRY RUN] No data will be written.');
  print('');

  // Parse CSV
  final file = File(csvPath);
  if (!file.existsSync()) {
    print('ERROR: CSV file not found: $csvPath');
    exit(1);
  }

  final records = _parseCsv(file.readAsStringSync());

  final makers = records.where((r) => r['type'] == 'maker').toList();
  final models = records.where((r) => r['type'] == 'model').toList();
  final grades = records.where((r) => r['type'] == 'grade').toList();

  print('Parsed records:');
  print('  Makers: ${makers.length}');
  print('  Models: ${models.length}');
  print('  Grades: ${grades.length}');
  print('');

  if (isDryRun) {
    print('[DRY RUN] Makers:');
    for (final m in makers) {
      print('  ${m['id']} | ${m['name']} (${m['name_en']}) | country: ${m['country']} | order: ${m['display_order']}');
    }
    print('');
    print('[DRY RUN] Models (first 10):');
    for (final m in models.take(10)) {
      print('  ${m['id']} | ${m['name']} | maker: ${m['parent_id']} | body: ${m['body_type']} | from: ${m['production_start_year']}');
    }
    print('  ... (${models.length} total)');
    print('');
    print('[DRY RUN] Grades:');
    for (final g in grades) {
      print('  ${g['id']} | ${g['name']}');
    }
    print('');
    print('[DRY RUN] Complete. No changes made.');
    return;
  }

  // Generate seed script
  _generateSeedFile(makers, models, grades, doClear);

  print('');
  print('Generated: scripts/_seed_vehicle_master_generated.dart');
  print('');
  print('Next step: Run the app and trigger seed from debug menu,');
  print('OR run the generated integration test:');
  print('  flutter test scripts/_seed_vehicle_master_generated.dart');
  print('');
  print('TIP: VehicleMasterService.seedMasterData() reads from VehicleMasterData');
  print('     (lib/data/vehicle_master_data.dart) and writes to Firestore.');
  print('     Update that file AND this CSV together when adding new models.');
}

/// Parse CSV, skip comment lines and blank lines
List<Map<String, String>> _parseCsv(String content) {
  final lines = content.split('\n');
  final records = <Map<String, String>>[];

  String? headerLine;
  List<String> headers = [];

  for (var line in lines) {
    line = line.trim();
    if (line.isEmpty || line.startsWith('#')) continue;

    if (headerLine == null) {
      headerLine = line;
      headers = line.split(',').map((h) => h.trim()).toList();
      continue;
    }

    // Skip comment-only data lines
    final withoutInlineComment = line.contains('#')
        ? line.substring(0, line.indexOf('#')).trim()
        : line;
    if (withoutInlineComment.isEmpty) continue;

    final values = withoutInlineComment.split(',').map((v) => v.trim()).toList();
    if (values.length < headers.length) {
      // Pad with empty strings
      while (values.length < headers.length) {
        values.add('');
      }
    }

    final record = <String, String>{};
    for (var i = 0; i < headers.length; i++) {
      record[headers[i]] = i < values.length ? values[i] : '';
    }

    if (record['type']?.isNotEmpty == true && record['id']?.isNotEmpty == true) {
      records.add(record);
    }
  }

  return records;
}

void _generateSeedFile(
  List<Map<String, String>> makers,
  List<Map<String, String>> models,
  List<Map<String, String>> grades,
  bool doClear,
) {
  final buf = StringBuffer();
  buf.writeln('// AUTO-GENERATED by scripts/import_vehicle_master.dart');
  buf.writeln('// Generated at: ${DateTime.now().toIso8601String()}');
  buf.writeln('// DO NOT EDIT MANUALLY — regenerate from CSV instead');
  buf.writeln('');
  buf.writeln('// To apply this seed:');
  buf.writeln('// 1. Ensure Firebase emulator is running (or use production with caution)');
  buf.writeln('// 2. Call VehicleMasterSeeder.seed() from your app or test');
  buf.writeln('');
  buf.writeln("import 'package:cloud_firestore/cloud_firestore.dart';");
  buf.writeln('');
  buf.writeln('class VehicleMasterSeeder {');
  buf.writeln('  VehicleMasterSeeder._();');
  buf.writeln('');
  buf.writeln('  static Future<void> seed() async {');
  buf.writeln('    final db = FirebaseFirestore.instance;');
  if (doClear) {
    buf.writeln('    await _clearCollections(db);');
  }
  buf.writeln('    final batch = db.batch();');
  buf.writeln('');

  buf.writeln('    // MAKERS');
  for (final m in makers) {
    buf.writeln("    batch.set(db.collection('vehicle_makers').doc('${m['id']}'), {");
    buf.writeln("      'name': '${m['name']}',");
    buf.writeln("      'nameEn': '${m['name_en']}',");
    buf.writeln("      'country': '${m['country']}',");
    buf.writeln("      'displayOrder': ${(m['display_order']?.isEmpty ?? true) ? '100' : m['display_order']},");
    buf.writeln("      'isActive': true,");
    buf.writeln("    });");
  }

  buf.writeln('');
  buf.writeln('    // MODELS');
  for (final m in models) {
    buf.writeln("    batch.set(db.collection('vehicle_models').doc('${m['id']}'), {");
    buf.writeln("      'makerId': '${m['parent_id']}',");
    buf.writeln("      'name': '${m['name']}',");
    buf.writeln("      'nameEn': '${m['name_en']}',");
    if (m['body_type']?.isNotEmpty == true) {
      buf.writeln("      'bodyType': '${m['body_type']}',");
    }
    if (m['production_start_year']?.isNotEmpty == true) {
      buf.writeln("      'productionStartYear': ${m['production_start_year']},");
    }
    if (m['production_end_year']?.isNotEmpty == true) {
      buf.writeln("      'productionEndYear': ${m['production_end_year']},");
    }
    buf.writeln("      'displayOrder': ${(m['display_order']?.isEmpty ?? true) ? '100' : m['display_order']},");
    buf.writeln("      'isActive': true,");
    buf.writeln("    });");
  }

  buf.writeln('');
  buf.writeln('    await batch.commit();');
  buf.writeln('    print(\'Seeded ${makers.length} makers, ${models.length} models\');');
  buf.writeln('  }');

  if (doClear) {
    buf.writeln('');
    buf.writeln('  static Future<void> _clearCollections(FirebaseFirestore db) async {');
    buf.writeln("    final collections = ['vehicle_makers', 'vehicle_models', 'vehicle_grades'];");
    buf.writeln('    for (final col in collections) {');
    buf.writeln('      final docs = await db.collection(col).get();');
    buf.writeln('      final batch = db.batch();');
    buf.writeln('      for (final doc in docs.docs) { batch.delete(doc.reference); }');
    buf.writeln('      await batch.commit();');
    buf.writeln('    }');
    buf.writeln('  }');
  }

  buf.writeln('}');

  File('scripts/_seed_vehicle_master_generated.dart').writeAsStringSync(buf.toString());
}

String? _getArg(List<String> args, String flag) {
  final idx = args.indexOf(flag);
  if (idx == -1 || idx + 1 >= args.length) return null;
  return args[idx + 1];
}

File: scripts/import_vehicle_master.js
#!/usr/bin/env node
/**
 * Vehicle Master Import — CSV → Firestore (Node.js / firebase-admin)
 *
 * `data/vehicle_masters.csv`（メーカー/車種/グレードのマスタ）を、アプリの
 * `VehicleMasterService` が読むネスト構造へ投入する。
 *
 *   vehicle_masters/makers/items/{id}
 *   vehicle_masters/models/items/{id}
 *
 * 既存の `import_vehicle_master.dart` は Dart コードを生成するだけで Firestore に
 * 書き込めず、しかもトップレベル `vehicle_makers`/`vehicle_models` を想定していて
 * サービスが読むネスト構造と不一致だった。本スクリプトはその不整合を解消する。
 *
 * Usage:
 *   node scripts/import_vehicle_master.js [--dry-run] [--emulator]
 *
 * Options:
 *   --dry-run    Firestore に書かず、投入予定を標準出力に表示する
 *   --emulator   Firebase Emulator (localhost:8080) に接続する
 *
 * Requirements:
 *   cd scripts && npm install   # firebase-admin
 *
 * Example:
 *   firebase emulators:start --only firestore
 *   node scripts/import_vehicle_master.js --dry-run
 *   node scripts/import_vehicle_master.js --emulator
 *
 *   # 本番（要人手承認・vehicle_masters は rules で write=false のため Admin SDK 必須）
 *   export GOOGLE_APPLICATION_CREDENTIALS=path/to/serviceAccount.json
 *   node scripts/import_vehicle_master.js
 *
 * NOTE (grades):
 *   CSV の grade 行は parent_id を持たない「共通グレード」（S/G/X/Z…）で、
 *   `VehicleMasterService.getGradesForModel` は `where('modelId', ...)` で絞り込む
 *   ため、modelId 無しの共通グレードは Firestore からは供給できない。よって grades は
 *   アプリ側の共通グレード・フォールバック（VehicleMasterData.getCommonGrades）に委ね、
 *   本スクリプトでは makers / models のみ投入する（=88車種への拡充。grades は非回帰）。
 *
 * NOTE (本番インデックス):
 *   本番で Firestore データが実際に使われるには、`items` コレクショングループの
 *   複合インデックスが必要（未デプロイ時はクエリが失敗し静的フォールバックに落ちる）。
 *   firestore.indexes.json に追加済み。`firebase deploy --only firestore:indexes` は人手タスク。
 */

const fs = require('fs');
const path = require('path');

const isDryRun = process.argv.includes('--dry-run');
const useEmulator = process.argv.includes('--emulator');

if (useEmulator) {
  process.env.FIRESTORE_EMULATOR_HOST =
    process.env.FIRESTORE_EMULATOR_HOST || 'localhost:8080';
}

const admin = (() => {
  try {
    return require('firebase-admin');
  } catch {
    console.error('[ERROR] firebase-admin が見つかりません。');
    console.error('        cd scripts && npm install を実行してください。');
    process.exit(1);
  }
})();

if (!admin.apps.length) {
  if (useEmulator) {
    admin.initializeApp({ projectId: 'trust-car-platform' });
  } else {
    admin.initializeApp({ credential: admin.credential.applicationDefault() });
  }
}

const db = admin.firestore();
db.settings({ ignoreUndefinedProperties: true });

// ---------------------------------------------------------------------------
// CSV パース
// ---------------------------------------------------------------------------
const CSV_PATH = path.join(__dirname, '..', 'data', 'vehicle_masters.csv');

/** 数値化（空文字/未定義は undefined＝書き込まない） */
function num(v) {
  if (v === undefined || v === null || v.trim() === '') return undefined;
  const n = parseInt(v.trim(), 10);
  return Number.isNaN(n) ? undefined : n;
}

/** 空文字は undefined（書き込まない）に正規化 */
function str(v) {
  if (v === undefined || v === null) return undefined;
  const s = v.trim();
  return s === '' ? undefined : s;
}

function parseCsv() {
  const raw = fs.readFileSync(CSV_PATH, 'utf8');
  const lines = raw.split(/\r?\n/);
  const makers = [];
  const models = [];
  let gradeCount = 0;

  // 列: type,id,parent_id,name,name_en,body_type,
  //     production_start_year,production_end_year,display_order,country
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed === '' || trimmed.startsWith('#')) continue;
    const cols = line.split(',');
    const type = (cols[0] || '').trim();
    if (type === 'type') continue; // ヘッダ行

    const id = str(cols[1]);
    const parentId = str(cols[2]);
    const name = str(cols[3]);
    const nameEn = str(cols[4]);
    const bodyType = str(cols[5]);
    const startYear = num(cols[6]);
    const endYear = num(cols[7]);
    const displayOrder = num(cols[8]) ?? 0;
    const country = str(cols[9]);

    if (type === 'maker') {
      if (!id || !name) continue;
      makers.push({
        id,
        data: {
          name,
          nameEn: nameEn ?? name,
          country: country ?? 'JP',
          displayOrder,
          isActive: true,
        },
      });
    } else if (type === 'model') {
      if (!id || !parentId || !name) continue;
      models.push({
        id,
        data: {
          makerId: parentId,
          name,
          nameEn, // 任意（未設定なら undefined→書き込まない）
          bodyType, // BodyType.fromString がパース。未設定なら undefined
          productionStartYear: startYear,
          productionEndYear: endYear,
          displayOrder,
          isActive: true,
        },
      });
    } else if (type === 'grade') {
      // 共通グレード（modelId 無し）はアプリのフォールバックに委ねる（NOTE参照）。
      gradeCount++;
    }
  }

  return { makers, models, gradeCount };
}

// ---------------------------------------------------------------------------
// 実行
// ---------------------------------------------------------------------------
async function main() {
  console.log('=== Vehicle Master Import ===');
  console.log(`dry-run  : ${isDryRun}`);
  console.log(`emulator : ${useEmulator}`);
  console.log(`csv      : ${CSV_PATH}`);
  console.log('');

  const { makers, models, gradeCount } = parseCsv();

  console.log('投入サマリー:');
  console.log(`  vehicle_masters/makers/items : ${makers.length}`);
  console.log(`  vehicle_masters/models/items : ${models.length}`);
  console.log(
    `  grades（共通・フォールバックに委ね未投入）: ${gradeCount}`,
  );
  console.log('');

  if (isDryRun) {
    console.log('--- [DRY RUN] makers ---');
    for (const m of makers) {
      console.log(`  ${m.id} — ${m.data.name} (${m.data.nameEn})`);
    }
    console.log('--- [DRY RUN] models（先頭20件） ---');
    for (const m of models.slice(0, 20)) {
      console.log(
        `  ${m.id} — ${m.data.name} [maker=${m.data.makerId}, body=${
          m.data.bodyType ?? '-'
        }]`,
      );
    }
    if (models.length > 20) console.log(`  ...（他 ${models.length - 20} 件）`);
    console.log('\n--- [DRY RUN] 完了（Firestore への書き込みは行っていません）---');
    return;
  }

  const makersCol = db
    .collection('vehicle_masters')
    .doc('makers')
    .collection('items');
  const modelsCol = db
    .collection('vehicle_masters')
    .doc('models')
    .collection('items');

  const batch = db.batch();
  for (const m of makers) {
    batch.set(makersCol.doc(m.id), m.data, { merge: true });
  }
  for (const m of models) {
    batch.set(modelsCol.doc(m.id), m.data, { merge: true });
  }
  await batch.commit();

  console.log(
    `[SUCCESS] makers ${makers.length} 件 / models ${models.length} 件を Firestore に投入しました。`,
  );
  console.log(
    'grades は共通グレードのため未投入（アプリの VehicleMasterData.getCommonGrades にフォールバック）。',
  );
}

main().catch((err) => {
  console.error('[ERROR]', err);
  process.exit(1);
});
